File: engine/sdks/typescript/envoy-client/src/stringify.ts
import type * as protocol from "@rivetkit/engine-envoy-protocol";
import { idToStr } from "./utils";

function stringifyArrayBuffer(buffer: ArrayBuffer): string {
	return `ArrayBuffer(${buffer.byteLength})`;
}

function stringifyBigInt(value: bigint): string {
	return `${value}n`;
}

function stringifyMap(map: ReadonlyMap<string, string>): string {
	const entries = Array.from(map.entries())
		.map(([k, v]) => `"${k}": "${v}"`)
		.join(", ");
	return `Map(${map.size}){${entries}}`;
}

function stringifyMessageId(messageId: protocol.MessageId): string {
	return `MessageId{gatewayId: ${idToStr(messageId.gatewayId)}, requestId: ${idToStr(messageId.requestId)}, messageIndex: ${messageId.messageIndex}}`;
}

export function stringifyToRivetTunnelMessageKind(
	kind: protocol.ToRivetTunnelMessageKind,
): string {
	switch (kind.tag) {
		case "ToRivetResponseStart": {
			const { status, headers, body, stream } = kind.val;
			const bodyStr = body === null ? "null" : stringifyArrayBuffer(body);
			return `ToRivetResponseStart{status: ${status}, headers: ${stringifyMap(headers)}, body: ${bodyStr}, stream: ${stream}}`;
		}
		case "ToRivetResponseChunk": {
			const { body, finish } = kind.val;
			return `ToRivetResponseChunk{body: ${stringifyArrayBuffer(body)}, finish: ${finish}}`;
		}
		case "ToRivetResponseAbort":
			return "ToRivetResponseAbort";
		case "ToRivetWebSocketOpen": {
			const { canHibernate } = kind.val;
			return `ToRivetWebSocketOpen{canHibernate: ${canHibernate}}`;
		}
		case "ToRivetWebSocketMessage": {
			const { data, binary } = kind.val;
			return `ToRivetWebSocketMessage{data: ${stringifyArrayBuffer(data)}, binary: ${binary}}`;
		}
		case "ToRivetWebSocketMessageAck": {
			const { index } = kind.val;
			return `ToRivetWebSocketMessageAck{index: ${index}}`;
		}
		case "ToRivetWebSocketClose": {
			const { code, reason, hibernate } = kind.val;
			const codeStr = code === null ? "null" : code.toString();
			const reasonStr = reason === null ? "null" : `"${reason}"`;
			return `ToRivetWebSocketClose{code: ${codeStr}, reason: ${reasonStr}, hibernate: ${hibernate}}`;
		}
	}
}

export function stringifyToEnvoyTunnelMessageKind(
	kind: protocol.ToEnvoyTunnelMessageKind,
): string {
	switch (kind.tag) {
		case "ToEnvoyRequestStart": {
			const { actorId, method, path, headers, body, stream } = kind.val;
			const bodyStr = body === null ? "null" : stringifyArrayBuffer(body);
			return `ToEnvoyRequestStart{actorId: "${actorId}", method: "${method}", path: "${path}", headers: ${stringifyMap(headers)}, body: ${bodyStr}, stream: ${stream}}`;
		}
		case "ToEnvoyRequestChunk": {
			const { body, finish } = kind.val;
			return `ToEnvoyRequestChunk{body: ${stringifyArrayBuffer(body)}, finish: ${finish}}`;
		}
		case "ToEnvoyRequestAbort":
			return "ToEnvoyRequestAbort";
		case "ToEnvoyWebSocketOpen": {
			const { actorId, path, headers } = kind.val;
			return `ToEnvoyWebSocketOpen{actorId: "${actorId}", path: "${path}", headers: ${stringifyMap(headers)}}`;
		}
		case "ToEnvoyWebSocketMessage": {
			const { data, binary } = kind.val;
			return `ToEnvoyWebSocketMessage{data: ${stringifyArrayBuffer(data)}, binary: ${binary}}`;
		}
		case "ToEnvoyWebSocketClose": {
			const { code, reason } = kind.val;
			const codeStr = code === null ? "null" : code.toString();
			const reasonStr = reason === null ? "null" : `"${reason}"`;
			return `ToEnvoyWebSocketClose{code: ${codeStr}, reason: ${reasonStr}}`;
		}
	}
}

export function stringifyCommand(command: protocol.Command): string {
	switch (command.tag) {
		case "CommandStartActor": {
			const { config, hibernatingRequests } = command.val;
			const keyStr = config.key === null ? "null" : `"${config.key}"`;
			const inputStr =
				config.input === null
					? "null"
					: stringifyArrayBuffer(config.input);
			const hibernatingRequestsStr =
				hibernatingRequests.length > 0
					? `[${hibernatingRequests.map((hr) => `{gatewayId: ${idToStr(hr.gatewayId)}, requestId: ${idToStr(hr.requestId)}}`).join(", ")}]`
					: "[]";
			return `CommandStartActor{config: {name: "${config.name}", key: ${keyStr}, createTs: ${stringifyBigInt(config.createTs)}, input: ${inputStr}}, hibernatingRequests: ${hibernatingRequestsStr}}`;
		}
		case "CommandStopActor": {
			const { reason } = command.val;
			return `CommandStopActor{reason: ${reason}}`;
		}
	}
}

export function stringifyCommandWrapper(
	wrapper: protocol.CommandWrapper,
): string {
	return `CommandWrapper{actorId: "${wrapper.checkpoint.actorId}", generation: "${wrapper.checkpoint.generation}", index: ${stringifyBigInt(wrapper.checkpoint.index)}, inner: ${stringifyCommand(wrapper.inner)}}`;
}

export function stringifyEvent(event: protocol.Event): string {
	switch (event.tag) {
		case "EventActorIntent": {
			const { intent } = event.val;
			const intentStr =
				intent.tag === "ActorIntentSleep"
					? "Sleep"
					: intent.tag === "ActorIntentStop"
						? "Stop"
						: "Unknown";
			return `EventActorIntent{intent: ${intentStr}}`;
		}
		case "EventActorStateUpdate": {
			const { state } = event.val;
			let stateStr: string;
			if (state.tag === "ActorStateRunning") {
				stateStr = "Running";
			} else if (state.tag === "ActorStateStopped") {
				const { code, message } = state.val;
				const messageStr = message === null ? "null" : `"${message}"`;
				stateStr = `Stopped{code: ${code}, message: ${messageStr}}`;
			} else {
				stateStr = "Unknown";
			}
			return `EventActorStateUpdate{state: ${stateStr}}`;
		}
		case "EventActorSetAlarm": {
			const { alarmTs } = event.val;
			const alarmTsStr =
				alarmTs === null ? "null" : stringifyBigInt(alarmTs);
			return `EventActorSetAlarm{alarmTs: ${alarmTsStr}}`;
		}
	}
}

export function stringifyEventWrapper(wrapper: protocol.EventWrapper): string {
	return `EventWrapper{actorId: ${wrapper.checkpoint.actorId}, generation: "${wrapper.checkpoint.generation}", index: ${stringifyBigInt(wrapper.checkpoint.index)}, inner: ${stringifyEvent(wrapper.inner)}}`;
}

export function stringifyToRivet(message: protocol.ToRivet): string {
	switch (message.tag) {
		case "ToRivetInit": {
			const { envoyKey, version, prepopulateActorNames, metadata } =
				message.val;
			const prepopulateActorNamesStr =
				prepopulateActorNames === null
					? "null"
					: `Map(${prepopulateActorNames.size})`;
			const metadataStr = metadata === null ? "null" : `"${metadata}"`;
			return `ToRivetInit{envoyKey: "${envoyKey}", version: ${version}, prepopulateActorNames: ${prepopulateActorNamesStr}, metadata: ${metadataStr}}`;
		}
		case "ToRivetEvents": {
			const events = message.val;
			return `ToRivetEvents{count: ${events.length}, events: [${events.map((e) => stringifyEventWrapper(e)).join(", ")}]}`;
		}
		case "ToRivetAckCommands": {
			const { lastCommandCheckpoints } = message.val;
			const checkpointsStr =
				lastCommandCheckpoints.length > 0
					? `[${lastCommandCheckpoints.map((cp) => `{actorId: "${cp.actorId}", index: ${stringifyBigInt(cp.index)}}`).join(", ")}]`
					: "[]";
			return `ToRivetAckCommands{lastCommandCheckpoints: ${checkpointsStr}}`;
		}
		case "ToRivetStopping":
			return "ToRivetStopping";
		case "ToRivetPong": {
			const { ts } = message.val;
			return `ToRivetPong{ts: ${stringifyBigInt(ts)}}`;
		}
		case "ToRivetKvRequest": {
			const { actorId, requestId, data } = message.val;
			const dataStr = stringifyKvRequestData(data);
			return `ToRivetKvRequest{actorId: "${actorId}", requestId: ${requestId}, data: ${dataStr}}`;
		}
		case "ToRivetTunnelMessage": {
			const { messageId, messageKind } = message.val;
			return `ToRivetTunnelMessage{messageId: ${stringifyMessageId(messageId)}, messageKind: ${stringifyToRivetTunnelMessageKind(messageKind)}}`;
		}
	}
}

export function stringifyToEnvoy(message: protocol.ToEnvoy): string {
	switch (message.tag) {
		case "ToEnvoyInit": {
			const { metadata } = message.val;
			const metadataStr = `{envoyLostThreshold: ${stringifyBigInt(metadata.envoyLostThreshold)}, actorStopThreshold: ${stringifyBigInt(metadata.actorStopThreshold)}, serverlessDrainGracePeriod: ${metadata.serverlessDrainGracePeriod === null ? "null" : stringifyBigInt(metadata.serverlessDrainGracePeriod)}, maxResponsePayloadSize: ${stringifyBigInt(metadata.maxResponsePayloadSize)}}`;
			return `ToEnvoyInit{metadata: ${metadataStr}}`;
		}
		case "ToEnvoyCommands": {
			const commands = message.val;
			return `ToEnvoyCommands{count: ${commands.length}, commands: [${commands.map((c) => stringifyCommandWrapper(c)).join(", ")}]}`;
		}
		case "ToEnvoyAckEvents": {
			const { lastEventCheckpoints } = message.val;
			const checkpointsStr =
				lastEventCheckpoints.length > 0
					? `[${lastEventCheckpoints.map((cp) => `{actorId: "${cp.actorId}", index: ${stringifyBigInt(cp.index)}}`).join(", ")}]`
					: "[]";
			return `ToEnvoyAckEvents{lastEventCheckpoints: ${checkpointsStr}}`;
		}
		case "ToEnvoyKvResponse": {
			const { requestId, data } = message.val;
			const dataStr = stringifyKvResponseData(data);
			return `ToEnvoyKvResponse{requestId: ${requestId}, data: ${dataStr}}`;
		}
		case "ToEnvoyTunnelMessage": {
			const { messageId, messageKind } = message.val;
			return `ToEnvoyTunnelMessage{messageId: ${stringifyMessageId(messageId)}, messageKind: ${stringifyToEnvoyTunnelMessageKind(messageKind)}}`;
		}
		case "ToEnvoyPing": {
			const { ts } = message.val;
			return `ToEnvoyPing{ts: ${stringifyBigInt(ts)}}`;
		}
	}
}

function stringifyKvRequestData(data: protocol.KvRequestData): string {
	switch (data.tag) {
		case "KvGetRequest": {
			const { keys } = data.val;
			return `KvGetRequest{keys: ${keys.length}}`;
		}
		case "KvListRequest": {
			const { query, reverse, limit } = data.val;
			const reverseStr = reverse === null ? "null" : reverse.toString();
			const limitStr = limit === null ? "null" : stringifyBigInt(limit);
			return `KvListRequest{query: ${stringifyKvListQuery(query)}, reverse: ${reverseStr}, limit: ${limitStr}}`;
		}
		case "KvPutRequest": {
			const { keys, values } = data.val;
			return `KvPutRequest{keys: ${keys.length}, values: ${values.length}}`;
		}
		case "KvDeleteRequest": {
			const { keys } = data.val;
			return `KvDeleteRequest{keys: ${keys.length}}`;
		}
		case "KvDeleteRangeRequest": {
			const { start, end } = data.val;
			return `KvDeleteRangeRequest{start: ${stringifyArrayBuffer(start)}, end: ${stringifyArrayBuffer(end)}}`;
		}
		case "KvDropRequest":
			return "KvDropRequest";
	}
}

function stringifyKvListQuery(query: protocol.KvListQuery): string {
	switch (query.tag) {
		case "KvListAllQuery":
			return "KvListAllQuery";
		case "KvListRangeQuery": {
			const { start, end, exclusive } = query.val;
			return `KvListRangeQuery{start: ${stringifyArrayBuffer(start)}, end: ${stringifyArrayBuffer(end)}, exclusive: ${exclusive}}`;
		}
		case "KvListPrefixQuery": {
			const { key } = query.val;
			return `KvListPrefixQuery{key: ${stringifyArrayBuffer(key)}}`;
		}
	}
}

function stringifyKvResponseData(data: protocol.KvResponseData): string {
	switch (data.tag) {
		case "KvErrorResponse": {
			const { message } = data.val;
			return `KvErrorResponse{message: "${message}"}`;
		}
		case "KvGetResponse": {
			const { keys, values, metadata } = data.val;
			return `KvGetResponse{keys: ${keys.length}, values: ${values.length}, metadata: ${metadata.length}}`;
		}
		case "KvListResponse": {
			const { keys, values, metadata } = data.val;
			return `KvListResponse{keys: ${keys.length}, values: ${values.length}, metadata: ${metadata.length}}`;
		}
		case "KvPutResponse":
			return "KvPutResponse";
		case "KvDeleteResponse":
			return "KvDeleteResponse";
		case "KvDropResponse":
			return "KvDropResponse";
	}
}

File: engine/sdks/rust/envoy-client/src/stringify.rs
use rivet_envoy_protocol as protocol;
use rivet_util_serde::HashableMap;

use crate::utils::id_to_str;

fn stringify_bytes(data: &[u8]) -> String {
	format!("Bytes({})", data.len())
}

fn stringify_map(map: &HashableMap<String, String>) -> String {
	let entries: Vec<String> = map
		.iter()
		.map(|(k, v)| format!("\"{k}\": \"{v}\""))
		.collect();
	format!("Map({}){{{}}}", map.len(), entries.join(", "))
}

fn stringify_message_id(msg_id: &protocol::MessageId) -> String {
	format!(
		"MessageId{{gatewayId: {}, requestId: {}, messageIndex: {}}}",
		id_to_str(&msg_id.gateway_id),
		id_to_str(&msg_id.request_id),
		msg_id.message_index
	)
}

pub fn stringify_to_rivet_tunnel_message_kind(kind: &protocol::ToRivetTunnelMessageKind) -> String {
	match kind {
		protocol::ToRivetTunnelMessageKind::ToRivetResponseStart(val) => {
			let body_str = match &val.body {
				Some(b) => stringify_bytes(b),
				None => "null".to_string(),
			};
			format!(
				"ToRivetResponseStart{{status: {}, headers: {}, body: {}, stream: {}}}",
				val.status,
				stringify_map(&val.headers),
				body_str,
				val.stream
			)
		}
		protocol::ToRivetTunnelMessageKind::ToRivetResponseChunk(val) => {
			format!(
				"ToRivetResponseChunk{{body: {}, finish: {}}}",
				stringify_bytes(&val.body),
				val.finish
			)
		}
		protocol::ToRivetTunnelMessageKind::ToRivetResponseAbort => {
			"ToRivetResponseAbort".to_string()
		}
		protocol::ToRivetTunnelMessageKind::ToRivetWebSocketOpen(val) => {
			format!(
				"ToRivetWebSocketOpen{{canHibernate: {}}}",
				val.can_hibernate
			)
		}
		protocol::ToRivetTunnelMessageKind::ToRivetWebSocketMessage(val) => {
			format!(
				"ToRivetWebSocketMessage{{data: {}, binary: {}}}",
				stringify_bytes(&val.data),
				val.binary
			)
		}
		protocol::ToRivetTunnelMessageKind::ToRivetWebSocketMessageAck(val) => {
			format!("ToRivetWebSocketMessageAck{{index: {}}}", val.index)
		}
		protocol::ToRivetTunnelMessageKind::ToRivetWebSocketClose(val) => {
			let code_str = match &val.code {
				Some(c) => c.to_string(),
				None => "null".to_string(),
			};
			let reason_str = match &val.reason {
				Some(r) => format!("\"{r}\""),
				None => "null".to_string(),
			};
			format!(
				"ToRivetWebSocketClose{{code: {code_str}, reason: {reason_str}, hibernate: {}}}",
				val.hibernate
			)
		}
	}
}

pub fn stringify_to_envoy_tunnel_message_kind(kind: &protocol::ToEnvoyTunnelMessageKind) -> String {
	match kind {
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyRequestStart(val) => {
			let body_str = match &val.body {
				Some(b) => stringify_bytes(b),
				None => "null".to_string(),
			};
			format!(
				"ToEnvoyRequestStart{{actorId: \"{}\", method: \"{}\", path: \"{}\", headers: {}, body: {}, stream: {}}}",
				val.actor_id,
				val.method,
				val.path,
				stringify_map(&val.headers),
				body_str,
				val.stream
			)
		}
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyRequestChunk(val) => {
			format!(
				"ToEnvoyRequestChunk{{body: {}, finish: {}}}",
				stringify_bytes(&val.body),
				val.finish
			)
		}
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyRequestAbort => {
			"ToEnvoyRequestAbort".to_string()
		}
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyWebSocketOpen(val) => {
			format!(
				"ToEnvoyWebSocketOpen{{actorId: \"{}\", path: \"{}\", headers: {}}}",
				val.actor_id,
				val.path,
				stringify_map(&val.headers)
			)
		}
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyWebSocketMessage(val) => {
			format!(
				"ToEnvoyWebSocketMessage{{data: {}, binary: {}}}",
				stringify_bytes(&val.data),
				val.binary
			)
		}
		protocol::ToEnvoyTunnelMessageKind::ToEnvoyWebSocketClose(val) => {
			let code_str = match &val.code {
				Some(c) => c.to_string(),
				None => "null".to_string(),
			};
			let reason_str = match &val.reason {
				Some(r) => format!("\"{r}\""),
				None => "null".to_string(),
			};
			format!("ToEnvoyWebSocketClose{{code: {code_str}, reason: {reason_str}}}")
		}
	}
}

pub fn stringify_command(command: &protocol::Command) -> String {
	match command {
		protocol::Command::CommandStartActor(val) => {
			let key_str = match &val.config.key {
				Some(k) => format!("\"{k}\""),
				None => "null".to_string(),
			};
			let input_str = match &val.config.input {
				Some(i) => stringify_bytes(i),
				None => "null".to_string(),
			};
			let hib_str = if val.hibernating_requests.is_empty() {
				"[]".to_string()
			} else {
				let entries: Vec<String> = val
					.hibernating_requests
					.iter()
					.map(|hr| {
						format!(
							"{{gatewayId: {}, requestId: {}}}",
							id_to_str(&hr.gateway_id),
							id_to_str(&hr.request_id)
						)
					})
					.collect();
				format!("[{}]", entries.join(", "))
			};
			format!(
				"CommandStartActor{{config: {{name: \"{}\", key: {key_str}, createTs: {}, input: {input_str}}}, hibernatingRequests: {hib_str}}}",
				val.config.name, val.config.create_ts
			)
		}
		protocol::Command::CommandStopActor(val) => {
			format!("CommandStopActor{{reason: {:?}}}", val.reason)
		}
	}
}

pub fn stringify_command_wrapper(wrapper: &protocol::CommandWrapper) -> String {
	format!(
		"CommandWrapper{{actorId: \"{}\", generation: {}, index: {}, inner: {}}}",
		wrapper.checkpoint.actor_id,
		wrapper.checkpoint.generation,
		wrapper.checkpoint.index,
		stringify_command(&wrapper.inner)
	)
}

pub fn stringify_event(event: &protocol::Event) -> String {
	match event {
		protocol::Event::EventActorIntent(val) => {
			let intent_str = match &val.intent {
				protocol::ActorIntent::ActorIntentSleep => "Sleep",
				protocol::ActorIntent::ActorIntentStop => "Stop",
			};
			format!("EventActorIntent{{intent: {intent_str}}}")
		}
		protocol::Event::EventActorStateUpdate(val) => {
			let state_str = match &val.state {
				protocol::ActorState::ActorStateRunning => "Running".to_string(),
				protocol::ActorState::ActorStateStopped(stopped) => {
					let message_str = match &stopped.message {
						Some(m) => format!("\"{m}\""),
						None => "null".to_string(),
					};
					format!(
						"Stopped{{code: {:?}, message: {message_str}}}",
						stopped.code
					)
				}
			};
			format!("EventActorStateUpdate{{state: {state_str}}}")
		}
		protocol::Event::EventActorSetAlarm(val) => {
			let alarm_str = match val.alarm_ts {
				Some(ts) => ts.to_string(),
				None => "null".to_string(),
			};
			format!("EventActorSetAlarm{{alarmTs: {alarm_str}}}")
		}
	}
}

pub fn stringify_event_wrapper(wrapper: &protocol::EventWrapper) -> String {
	format!(
		"EventWrapper{{actorId: {}, generation: {}, index: {}, inner: {}}}",
		wrapper.checkpoint.actor_id,
		wrapper.checkpoint.generation,
		wrapper.checkpoint.index,
		stringify_event(&wrapper.inner)
	)
}

pub fn stringify_to_rivet(message: &protocol::ToRivet) -> String {
	match message {
		protocol::ToRivet::ToRivetMetadata(_) => "ToRivetMetadata".to_string(),
		protocol::ToRivet::ToRivetEvents(events) => {
			let event_strs: Vec<String> = events.iter().map(stringify_event_wrapper).collect();
			format!(
				"ToRivetEvents{{count: {}, events: [{}]}}",
				events.len(),
				event_strs.join(", ")
			)
		}
		protocol::ToRivet::ToRivetAckCommands(val) => {
			let checkpoints: Vec<String> = val
				.last_command_checkpoints
				.iter()
				.map(|cp| format!("{{actorId: \"{}\", index: {}}}", cp.actor_id, cp.index))
				.collect();
			format!(
				"ToRivetAckCommands{{lastCommandCheckpoints: [{}]}}",
				checkpoints.join(", ")
			)
		}
		protocol::ToRivet::ToRivetStopping => "ToRivetStopping".to_string(),
		protocol::ToRivet::ToRivetPong(val) => {
			format!("ToRivetPong{{ts: {}}}", val.ts)
		}
		protocol::ToRivet::ToRivetKvRequest(val) => {
			format!(
				"ToRivetKvRequest{{actorId: \"{}\", requestId: {}}}",
				val.actor_id, val.request_id
			)
		}
		protocol::ToRivet::ToRivetSqliteGetPagesRequest(val) => {
			format!(
				"ToRivetSqliteGetPagesRequest{{requestId: {}}}",
				val.request_id
			)
		}
		protocol::ToRivet::ToRivetSqliteCommitRequest(val) => {
			format!(
				"ToRivetSqliteCommitRequest{{requestId: {}}}",
				val.request_id
			)
		}
		protocol::ToRivet::ToRivetTunnelMessage(val) => {
			format!(
				"ToRivetTunnelMessage{{messageId: {}, messageKind: {}}}",
				stringify_message_id(&val.message_id),
				stringify_to_rivet_tunnel_message_kind(&val.message_kind)
			)
		}
	}
}

pub fn stringify_to_envoy(message: &protocol::ToEnvoy) -> String {
	match message {
		protocol::ToEnvoy::ToEnvoyInit(val) => {
			format!(
				"ToEnvoyInit{{metadata: {{envoyLostThreshold: {}, actorStopThreshold: {}}}}}",
				val.metadata.envoy_lost_threshold, val.metadata.actor_stop_threshold
			)
		}
		protocol::ToEnvoy::ToEnvoyCommands(commands) => {
			let cmd_strs: Vec<String> = commands.iter().map(stringify_command_wrapper).collect();
			format!(
				"ToEnvoyCommands{{count: {}, commands: [{}]}}",
				commands.len(),
				cmd_strs.join(", ")
			)
		}
		protocol::ToEnvoy::ToEnvoyAckEvents(val) => {
			let checkpoints: Vec<String> = val
				.last_event_checkpoints
				.iter()
				.map(|cp| format!("{{actorId: \"{}\", index: {}}}", cp.actor_id, cp.index))
				.collect();
			format!(
				"ToEnvoyAckEvents{{lastEventCheckpoints: [{}]}}",
				checkpoints.join(", ")
			)
		}
		protocol::ToEnvoy::ToEnvoyKvResponse(val) => {
			format!("ToEnvoyKvResponse{{requestId: {}}}", val.request_id)
		}
		protocol::ToEnvoy::ToEnvoySqliteGetPagesResponse(val) => {
			format!(
				"ToEnvoySqliteGetPagesResponse{{requestId: {}}}",
				val.request_id
			)
		}
		protocol::ToEnvoy::ToEnvoySqliteCommitResponse(val) => {
			format!(
				"ToEnvoySqliteCommitResponse{{requestId: {}}}",
				val.request_id
			)
		}
		protocol::ToEnvoy::ToEnvoyTunnelMessage(val) => {
			format!(
				"ToEnvoyTunnelMessage{{messageId: {}, messageKind: {}}}",
				stringify_message_id(&val.message_id),
				stringify_to_envoy_tunnel_message_kind(&val.message_kind)
			)
		}
		protocol::ToEnvoy::ToEnvoyPing(val) => {
			format!("ToEnvoyPing{{ts: {}}}", val.ts)
		}
	}
}
